---
title: Filtering of PBMC2 using COTAN
---

```{r}
library(dplyr)
library(COTAN)
library(Seurat)
library(tibble)
library(ggplot2)
library(zeallot)
library(DropletUtils)
options(parallelly.fork.enable = TRUE)

NAME = 'PBMC2'
DATASETS_FOLDER = '../dataset/'

inDataDir  = paste(DATASETS_FOLDER, NAME, sep='')
outDir = paste(DATASETS_FOLDER, NAME, '-Filtered/', sep='')
outDirRaw = paste(outDir, 'raw/', sep='')
out10XDir = paste(outDir, '10x/', sep='')
out10XRawDir = paste(outDirRaw, '10x/', sep='')
if (!dir.exists(outDir)) {
  dir.create(outDir, recursive = TRUE, showWarnings = FALSE)
}
if (!dir.exists(outDirRaw)) {
  dir.create(outDirRaw, recursive = TRUE, showWarnings = FALSE)
}

setLoggingLevel(2)
setLoggingFile(file.path(outDir, paste(NAME, "_analysis.log", sep="")))
```

Data loading

```{r}
dataset = Read10X(data.dir = inDataDir, strip.suffix = TRUE)
dataset = dataset[[1]]

sampleCondition <- NAME
PBMC2 <- COTAN(raw = dataset)
PBMC2 <- initializeMetaDataset(
  PBMC2,
  GEO = paste("10X ", NAME, sep=""),
  sequencingMethod = "10X",
  sampleCondition = sampleCondition
)
```

Inspect cellsâ€™ sizes

```{r}
cellSizePlot(PBMC2)
```

Drop cells with too many ritz reads as they are probably doublets

```{r}
cellsSizeThr <- 20000
PBMC2 <- addElementToMetaDataset(PBMC2, "Cells size threshold", cellsSizeThr)

cells_to_rem <- getCells(PBMC2)[getCellsSize(PBMC2) > cellsSizeThr]
PBMC2 <- dropGenesCells(PBMC2, cells = cells_to_rem)

cellSizePlot(PBMC2, splitPattern = "-", numCol = 2)
```

Inspect the number of expressed genes per cell

```{r}
genesSizePlot(PBMC2, splitPattern = "-", numCol = 2)
```

Drop cells with too high genes expession as they are probably doublets

```{r}
genesSizeThr <- 3500
PBMC2 <- addElementToMetaDataset(PBMC2, "Num genes threshold", genesSizeThr)

numExprGenes <- getNumExpressedGenes(PBMC2)
cells_to_rem <- names(numExprGenes)[numExprGenes > genesSizeThr]
PBMC2 <- dropGenesCells(PBMC2, cells = cells_to_rem)

genesSizePlot(PBMC2, splitPattern = "-", numCol = 2)
```

Check number of mithocondrial genes expressed in each cell

```{r}
mitGenesPattern <- "^[Mm][Tt]-"
getGenes(PBMC2)[grep(mitGenesPattern, getGenes(PBMC2))]
```

```{r}
c(mitPlot, mitSizes) %<-%
  mitochondrialPercentagePlot(PBMC2, genePrefix = mitGenesPattern,
                              splitPattern = "-", numCol = 2)

plot(mitPlot)
```

We drop cells with a too high percentage of mitocondrial genes (are likely dead)

```{r}
mitPercThr <- 10
PBMC2 <- addElementToMetaDataset(PBMC2, "Mitoc. perc. threshold", mitPercThr)

cells_to_rem <- rownames(mitSizes)[mitSizes[["mit.percentage"]] > mitPercThr]

PBMC2 <- dropGenesCells(PBMC2, cells = cells_to_rem)

c(mitPlot, mitSizes) %<-%
  mitochondrialPercentagePlot(PBMC2, genePrefix = mitGenesPattern,
                              splitPattern = "-", numCol = 2)

plot(mitPlot)
```

Check number of ribosomial genes expressed in each cell

```{r}
ribGenesPattern <- "^RP[SL]\\d+"
getGenes(PBMC2)[grep(ribGenesPattern, getGenes(PBMC2))]
```

```{r}
c(ribPlot, ribSizes) %<-%
  mitochondrialPercentagePlot(PBMC2, genePrefix = ribGenesPattern,
                              splitPattern = "-", numCol = 2)

plot(ribPlot)
```

Check no further outliers after all the culling

```{r}
cellSizePlot(PBMC2, splitPattern = "-", numCol = 2)
```

```{r}
genesSizePlot(PBMC2, splitPattern = "-", numCol = 2)
```

```{r}
if (!dir.exists(out10XRawDir)) {
  write10xCounts(out10XRawDir, getRawData(PBMC2))
}
```

Cleaning, round 1

```{r}
PBMC2 <- clean(PBMC2)

c(pcaCellsPlot, pcaCellsData, genesPlot, UDEPlot, nuPlot, zoomedNuPlot) %<-% cleanPlots(PBMC2)

plot(pcaCellsPlot)
```

```{r}
plot(genesPlot)
```

```{r}
PBMC2 <- addElementToMetaDataset(PBMC2, "Num drop B group", 0)
```

```{r}
plot(UDEPlot)
```

```{r}
plot(nuPlot)
```

```{r}
plot(zoomedNuPlot)  
```

Cleaning, round 2

```{r}
PBMC2 <- clean(PBMC2)

c(pcaCellsPlot, pcaCellsData, genesPlot, UDEPlot, nuPlot, zoomedNuPlot) %<-% cleanPlots(PBMC2)

plot(pcaCellsPlot)
```

```{r}
plot(pcaCellsData)
```

```{r}
plot(genesPlot)
```

```{r}
plot(UDEPlot)
```

```{r}
plot(nuPlot)
```

```{r}
plot(zoomedNuPlot)
```


```{r}
plot(cellSizePlot(PBMC2, splitPattern = "-", numCol = 2))
```

```{r}
plot(genesSizePlot(PBMC2, splitPattern = "-", numCol = 2))
```

Save the filtered dataset

```{r}
if (!dir.exists(out10XDir)) {
  write10xCounts(out10XDir, getRawData(PBMC2))
}
```

```{r}
saveRDS(PBMC2, file = file.path(outDir, paste0(sampleCondition, ".cotan.RDS")))
```

```{r}
sessionInfo()
```